cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/BufferingTable.java

```diff
@@ -565,13 +583,28 @@ public Scanner scan(byte[] startRow, byte[] stopRow) {
     } else {
       bufferMap = buff.subMap(startRow, true, stopRow, false);
     }
-    try {
-      return new BufferingScanner(bufferMap, scanPersisted(startRow, stopRow));
-    } catch (Exception e) {
-      LOG.debug("scan failed for table: " + getTransactionAwareName() +
-          ", start row: " + (startRow != null ? Bytes.toStringBinary(startRow) : "NULL") +
-          ", stop row: "  + (stopRow != null ? Bytes.toStringBinary(stopRow) : "NULL"), e);
-      throw new DataSetException("compareAndSwap failed", e);
+    bufferMap = applyFilter(bufferMap, scan.getFilter());
+    return bufferMap;
+  }
+
+  private NavigableMap<byte[], NavigableMap<byte[], Update>> applyFilter(
+                                                        NavigableMap<byte[], NavigableMap<byte[], Update>> bufferMap,
+                                                        @Nullable Object filter) {
+    if (filter == null) {
+      return bufferMap;
+    }
+
+    // todo: currently we support only FuzzyRowFilter as an experimental feature
+    if (filter instanceof FuzzyRowFilter) {
```

Comment:
The summary is that it is fine to start with support of one filter in this custom/special way. We'll learn from it and it will help to figure out the API for server side filters feature.